src/Content/Smilies.php

```diff
@@ -152,6 +153,129 @@ public static function getList(): array
 		return $params;
 	}
 
+	/**
+	 * Normalizes smiley shortcodes into texts with no special symbols.
+	 *
+	 * @return array
+	 *    'texts' => smilie shortcut
+	 *    'icons' => icon url or an empty string
+	 *    'norms' => normalized shortcut
+	 */
+	public static function getNormalizedList(): array
+	{
+		$smilies = self::getList();
+		$norms = [];
+		$icons = $smilies['icons'];
+		foreach ($smilies['texts'] as $i => $shortcode) {
+			// Extract urls
+			$icon = $icons[$i];
+			if (preg_match('/src="(.+?)"/', $icon, $match)) {
+				$icon = $match[1];
+			} else {
+				$icon = '';
+			}
+			$icons[$i] = $icon;
+
+			// Normalize name
+			$norm = preg_replace('/[\s\-:#~]/', '', $shortcode);
+			if (ctype_alnum($norm)) {
+				$norms[] = $norm;
+			} elseif (preg_match('#/smiley-(\w+)\.gif#', $icon, $match)) {
+				$norms[] = $match[1];
+			} else {
+				$norms[] = 'smiley' . $i;
+			}
+		}
+		$smilies['norms'] = $norms;
+		return $smilies;
+	}
+
+	/**
+	 * Finds all used smilies (denoted by quoting colons like :heart:) in the provided text and normalizes their usages.
+	 *
+	 * @param string $text that might contain smiley usages
+	 * @return array with smilie codes (colon included) as the keys, their image urls as values;
+	 *               the normalized string is put under the '' (empty string) key
+	 */
+	public static function extractUsedSmilies(string $text): array
+	{
+		$emojis = [];
+
+		$emojis[''] = BBCode::performWithEscapedTags($text, ['code'], function ($text) use (&$emojis) {
```

Comment:
How about given the array key an actual name like `'normalized'`?

Reply:
I've updated the code to use a param reference to pass the normalized string back.